src/celeritas/track/detail/TrackSortUtils.cu

```diff
@@ -68,6 +70,55 @@ void sort_impl(TrackSlots const& track_slots, F&& func, StreamId stream_id)
     CELER_DEVICE_CHECK_ERROR();
 }
 
+// PRE: action_accessor is sorted, i.e. i <= j ==> action_accessor(i) <=
+// action_accessor(j)
+template<class F>
+CELER_FUNCTION void tracks_per_action_impl(Span<ThreadId> offsets,
+                                           size_type size,
+                                           F&& action_accessor)
+{
+    ThreadId tid = celeritas::KernelParamCalculator::thread_id();
+
+    if ((tid < size) && tid.get() != 0)
+    {
+        ActionId current_action = action_accessor(tid);
+        ActionId previous_action = action_accessor(tid - 1);
+        if (current_action && current_action != previous_action)
+        {
+            offsets[current_action.get()] = tid;
+        }
+    }
+    // needed if the first action range has only one element
+    if (ActionId first; tid.get() == 0 && (first = action_accessor(tid)))
+    {
+        offsets[first.get()] = tid;
+    }
+}
+
+__global__ void tracks_per_action_kernel(DeviceRef<CoreStateData> const states,
+                                         Span<ThreadId> offsets,
+                                         size_type size,
+                                         TrackOrder order)
+{
```

Comment:
Yes this is why I wanted to launch a templated kernel. So you'd have a templated executor instead?

Reply:
Actually I'm not sure how an executor can help move that check to the host. Since executors are created on the host side and moved to device as a kernel argument we either have to: 

1. Use polymorphism and a base interface. Which doesn't work since objects are created on the host and the vtable will point to host function.
2. Have a generic `Executor<TrackOrder>` in which case the kernel should be generic as well.